linkerd/app/core/src/errors.rs

```diff
@@ -1,150 +1,187 @@
             }
         }
+
+        let status = http_status(error);
+        debug!(%status, "Handling error with HTTP response");
+        Ok(http::Response::builder()
+            .status(status)
+            .header(http::header::CONTENT_LENGTH, "0")
+            .body(B::default())
+            .expect("error response must be valid"))
     }
 }
 
-fn map_err_to_5xx(e: Error) -> StatusCode {
-    use crate::proxy::buffer;
-    use linkerd2_router::error as router;
-    use tower::load_shed::error as shed;
-
-    if let Some(ref c) = e.downcast_ref::<router::NoCapacity>() {
-        warn!("router at capacity ({})", c.0);
+fn http_status(error: Error) -> StatusCode {
+    if error.is::<router::NoCapacity>() {
         http::StatusCode::SERVICE_UNAVAILABLE
-    } else if let Some(_) = e.downcast_ref::<shed::Overloaded>() {
-        warn!("server overloaded, max-in-flight reached");
+    } else if error.is::<shed::Overloaded>() {
         http::StatusCode::SERVICE_UNAVAILABLE
-    } else if let Some(_) = e.downcast_ref::<buffer::Aborted>() {
-        warn!("request aborted because it reached the configured dispatch deadline");
+    } else if error.is::<buffer::Aborted>() {
         http::StatusCode::SERVICE_UNAVAILABLE
-    } else if let Some(_) = e.downcast_ref::<router::NotRecognized>() {
-        error!("could not recognize request");
-        http::StatusCode::BAD_GATEWAY
-    } else if let Some(err) = e.downcast_ref::<StatusError>() {
-        error!(%err.status, %err.message);
-        err.status
+    } else if error.is::<IdentityRequired>() {
+        http::StatusCode::FORBIDDEN
     } else {
-        // we probably should have handled this before?
-        error!("unexpected error: {}", e);
         http::StatusCode::BAD_GATEWAY
     }
 }
 
-impl std::fmt::Display for StatusError {
+fn set_grpc_status(error: Error, headers: &mut http::HeaderMap) -> Code {
+    const GRPC_STATUS: &'static str = "grpc-status";
+    const GRPC_MESSAGE: &'static str = "grpc-message";
+
+    if error.is::<router::NoCapacity>() {
+        let code = Code::Unavailable;
+        headers.insert(GRPC_STATUS, code_header(code));
+        headers.insert(
+            GRPC_MESSAGE,
+            HeaderValue::from_static("proxy router cache exhausted"),
+        );
+        code
+    } else if error.is::<shed::Overloaded>() {
+        let code = Code::Unavailable;
+        headers.insert(GRPC_STATUS, code_header(code));
+        headers.insert(
+            GRPC_MESSAGE,
+            HeaderValue::from_static("proxy max-concurrency exhausted"),
+        );
+        code
+    } else if error.is::<buffer::Aborted>() {
+        let code = Code::Unavailable;
+        headers.insert(GRPC_STATUS, code_header(code));
+        headers.insert(
+            GRPC_MESSAGE,
+            HeaderValue::from_static("proxy dispatch timed out"),
+        );
+        code
+    } else if error.is::<IdentityRequired>() {
+        let code = Code::FailedPrecondition;
```

Comment:
I don't think anything relied on the `Forbidden` status code, but is this changing the error type for required tap identity? Before in `require_identity_on_endpoint`, if the condition failed we returned an `error::StatusError` with a `Status = Forbidden`. Now it's a gRPC status code `FailedPrecondition`

Reply:
It's still a `FORBIDDEN` for HTTP requests; but is now a `FailedPrecondition` for gRPC requests.. reading through the descriptions i thought this was the closest ft.